File: backend/api/agents.py
"""
Agent API Router — Phase 3
Exposes endpoints for the RCA and Compliance LangGraph agents.
"""
import logging
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional

logger = logging.getLogger(__name__)

from agents.rca_agent import run_rca
from agents.compliance_agent import run_compliance_check
from agents.lessons_learned_job import run_lessons_learned_pipeline, get_all_insights

router = APIRouter(prefix="/api/agents", tags=["agents"])


# ── RCA ──────────────────────────────────────────────────────

class RCARequest(BaseModel):
    query: str
    asset_id: str

@router.post("/rca")
def rca_endpoint(req: RCARequest):
    """
    Run a Root Cause Analysis agent for a given failure question and asset.
    Returns ranked hypotheses with citations and recommended actions.
    """
    result = run_rca(req.query, req.asset_id)
    if result["status"] == "error":
        raise HTTPException(status_code=422, detail=result["message"])

    # Transform backend result to match frontend RCAResponse interface
    rca_data = result.get("rca", {})
    source_chunks = result.get("source_chunks", [])
    raw_hypotheses = rca_data.get("hypotheses", [])

    hypotheses = []
    for i, h in enumerate(raw_hypotheses):
        conf_label = h.get("confidence", "LOW").upper()
        if conf_label not in ("HIGH", "MEDIUM", "LOW"):
            conf_label = "MEDIUM"
        conf_pct = {"HIGH": 85, "MEDIUM": 60, "LOW": 30}.get(conf_label, 50)

        # Build citations from source_chunks
        citations = []
        for cite_ref in h.get("citations", []):
            parts = str(cite_ref).split(":")
            doc_id = parts[0] if parts else cite_ref
            page = int(parts[1]) if len(parts) > 1 and parts[1].isdigit() else 1
            # Find matching source chunk for score
            score = 0.0
            text_preview = ""
            for sc in source_chunks:
                if sc.get("doc_id") == doc_id:
                    score = sc.get("score", 0.0)
                    break
            citations.append({
                "doc_id": doc_id,
                "page": page,
                "score": score,
                "text_preview": text_preview,
            })

        hypotheses.append({
            "id": f"HYP-{i+1:03d}",
            "title": h.get("hypothesis", f"Hypothesis {i+1}"),
            "description": "; ".join(h.get("evidence", [])),
            "confidence": conf_pct,
            "confidence_label": conf_label,
            "citations": citations,
        })

    return {
        "asset_id": result.get("asset_id", req.asset_id),
        "query": result.get("query", req.query),
        "hypotheses": hypotheses,
        "graph_context": rca_data,
    }


# ── Compliance ───────────────────────────────────────────────

class ComplianceRequest(BaseModel):
    regulation_ids: List[str]
    plant_filter: Optional[str] = None

@router.post("/compliance")
def compliance_endpoint(req: ComplianceRequest):
    """
    Run a compliance gap analysis for a set of regulations.
    Returns per-regulation status, severity, and recommended actions.
    """
    result = run_compliance_check(req.regulation_ids, req.plant_filter)
    if result["status"] == "error":
        raise HTTPException(status_code=422, detail=result["message"])

    # Transform backend result to match frontend ComplianceResponse interface
    raw_gaps = result.get("gaps", [])
    gap_analysis = []
    for g in raw_gaps:
        status = g.get("status", "MISSING").upper()
        if status not in ("COMPLIANT", "GAP", "MISSING"):
            status = "GAP" if status == "UNKNOWN" else status
        severity = g.get("severity", "MEDIUM").upper()
        if severity not in ("LOW", "MEDIUM", "HIGH", "CRITICAL"):
            severity = "MEDIUM"

        evidence_docs = g.get("evidence_doc_ids", [])
        evidence_doc = evidence_docs[0] if evidence_docs else None

        gap_analysis.append({
            "regulation_id": g.get("regulation_id", ""),
            "clause": g.get("regulation_title", ""),
            "requirement": g.get("requirement_summary", g.get("gap_description", "")),
            "status": status,
            "evidence_doc": evidence_doc,
            "severity": severity,
            "recommended_action": g.get("recommended_action", "Review required"),
        })

    return {
        "plant_filter": req.plant_filter,
        "total_gaps": sum(1 for g in gap_analysis if g["status"] != "COMPLIANT"),
        "gap_analysis": gap_analysis,
    }


# ── Lessons Learned ──────────────────────────────────────────

class LessonsRequest(BaseModel):
    asset_type: Optional[str] = None

@router.post("/lessons-learned/run")
def run_lessons_endpoint(req: LessonsRequest = LessonsRequest()):
    """
    Trigger the lessons-learned batch pipeline manually.
    Scans recurring failures and writes insights to the graph.
    """
    result = run_lessons_learned_pipeline(req.asset_type)
    return result

@router.get("/lessons-learned/insights")
def get_insights_endpoint():
    """
    Retrieve all LessonsLearnedInsight nodes from Neo4j for the alerts dashboard.
    """
    return {"insights": get_all_insights()}

File: frontend/src/api/agents.ts
import { apiClient } from './client';

// Chat / Copilot
export interface ChatMessage {
  role: 'user' | 'assistant';
  content: string;
  citations?: Citation[];
}

export interface Citation {
  doc_id: string;
  page: number;
  score: number;
  text_preview: string;
}

export interface ChatResponse {
  answer: string;
  citations: Citation[];
  confidence: number;
}

export const sendChatMessage = async (query: string): Promise<ChatResponse> => {
  const res = await apiClient.post('/api/chat', { question: query });
  return res.data;
};

// RCA Agent
export interface RCAHypothesis {
  id: string;
  title: string;
  description: string;
  confidence: number;
  confidence_label: 'HIGH' | 'MEDIUM' | 'LOW';
  citations: Citation[];
}

export interface RCAResponse {
  asset_id: string;
  query: string;
  hypotheses: RCAHypothesis[];
  graph_context: Record<string, unknown>;
  error?: string;
}

export const runRcaAnalysis = async (assetId: string, query: string): Promise<RCAResponse> => {
  const res = await apiClient.post('/api/agents/rca', { asset_id: assetId, query });
  return res.data;
};

// Compliance Agent
export interface ComplianceGap {
  regulation_id: string;
  clause: string;
  requirement: string;
  status: 'COMPLIANT' | 'GAP' | 'MISSING';
  evidence_doc?: string;
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  recommended_action: string;
}

export interface ComplianceResponse {
  plant_filter?: string;
  total_gaps: number;
  gap_analysis: ComplianceGap[];
  error?: string;
}

export const runComplianceCheck = async (
  plantFilter?: string,
  regulationIds?: string[]
): Promise<ComplianceResponse> => {
  const res = await apiClient.post('/api/agents/compliance', {
    plant_filter: plantFilter,
    regulation_ids: regulationIds ?? [],
  });
  return res.data;
};

// Graph
export interface AssetGraph {
  asset_id: string;
  equipment: Record<string, unknown>;
  failure_modes: unknown[];
  work_orders: unknown[];
  incidents: unknown[];
}

export const getAssetGraph = async (assetId: string): Promise<AssetGraph> => {
  const res = await apiClient.get(`/api/graph/asset/${assetId}`);
  return res.data;
};

// Document Upload
export const uploadDocument = async (file: File, assetTag?: string): Promise<{ job_id: string }> => {
  const formData = new FormData();
  formData.append('file', file);
  if (assetTag) formData.append('asset_tag', assetTag);
  const res = await apiClient.post('/api/ingestion/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return res.data;
};

// Dashboard
export interface DashboardSummary {
  system_health_pct: number;
  services: { name: string; status: 'up' | 'down'; error?: string }[];
  document_count: number;
  equipment_count: number;
  failure_count: number;
  incident_count: number;
  open_work_order_count: number;
  non_conformance_count: number;
  critical_alerts: { asset_id: string; failure: string; severity: string; failure_id: string }[];
  pending_maintenance: { wo_id: string; title: string; priority: string; status: string; asset_id: string }[];
}

export const getDashboardSummary = async (): Promise<DashboardSummary> => {
  const res = await apiClient.get('/api/dashboard/summary');
  return res.data;
};

// Assets
export interface AssetSummary {
  asset_id: string;
  name: string;
  asset_type: string;
  plant: string;
  failure_count: number;
  work_order_count: number;
  incident_count: number;
}

export const getAssetsList = async (): Promise<{ assets: AssetSummary[]; total: number }> => {
  const res = await apiClient.get('/api/assets');
  return res.data;
};

// Reports
export interface ReportSummary {
  report_types: { id: string; title: string; description: string; icon: string; route: string }[];
  insights: any[];
  compliance_history: any[];
}

export const getReportsSummary = async (): Promise<ReportSummary> => {
  const res = await apiClient.get('/api/reports');
  return res.data;
};
